### ISSUE TITLE
求助

### ISSUE THREAD

**Issue by Flandern1211** [xurui-joei/DBCTNet#5]
作者大大，请问一下你在实现ConvTE模块的时候，有没有尝试过对Q也进行卷积操作？

**Comment by xurui-joei** [comment]
> 作者大大，请问一下你在实现ConvTE模块的时候，有没有尝试过对Q也进行卷积操作？

这个我好像没有试过

**Comment by machao0919** [comment]
作者大大，可以提供一下黄河三角洲的高光谱数据集资源吗

**Comment by xurui-joei** [comment]
你好，黄河三角洲的数据集可以关注一下彭江涛老师以及孙伟伟老师课题组的相关研究成果。HHK有很多不同时间节点，不同设备采集的数据集，我这边的HHK数据集是由彭老师那边帮忙提供的。




Zzz
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "xurui-joei/DBCTNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年8月12日(星期一) 晚上9:16
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [xurui-joei/DBCTNet] 求助 (Issue #5)





 
作者大大，可以提供一下黄河三角洲的高光谱数据集资源吗

**Comment by xurui-joei** [comment]
> 作者大大，可以提供一下黄河三角洲的高光谱数据集资源吗

你好，黄河三角洲的数据集可以关注一下彭江涛老师以及孙伟伟老师课题组的相关研究成果。HHK有很多不同时间节点，不同设备采集的数据集，我这边的HHK数据集是由彭老师那边帮忙提供的。

应该和下面这篇文章里用的HHK数据集是同一个原始数据版本
Q. Liu, J. Peng, N. Chen, W. Sun, Y. Ning and Q. Du, "Category-Specific Prototype Self-Refinement Contrastive Learning for Few-Shot Hyperspectral Image Classification," in IEEE Transactions on Geoscience and Remote Sensing, vol. 61, pp. 1-16, 2023.

**Comment by machao0919** [comment]
作者大大 请问一下您可以提供一下打印GPU Memory usage的这一小部分代码吗？

**Comment by xurui-joei** [comment]
> 作者大大 请问一下您可以提供一下打印GPU Memory usage的这一小部分代码吗？

你好。实验中显存占用使用nvidia-smi指令监控GPU，不同torch版本和硬件设备会导致显存占用发生变化

**Comment by NinJiaA** [comment]
作者大大您好，请问您计算Inference time是计算哪一部分的时间呢，我计算PU（参数一致）一个epoch的时间得到时间是0.81s，和论文差别较大。以下是代码
```python

```     for epoch in range(epochs):
        start_time = time.time()
        for phase in ['train','val']:
            running_loss = 0
            running_correct = 0
            for i, (data, target) in enumerate(data_loader[phase]):
                data, target = data.to(device), target.to(device)
                if phase == 'train':
                    model.train()
                    outputs = model(data)
                    loss = criterion(outputs, target)
                    optimizer.zero_grad()
                    loss.backward()
                    optimizer.step()
                    scheduler.step()
                else:
                    model.eval()
                    with torch.no_grad():
                        outputs = model(data)
                        loss = criterion(outputs, target)
                running_loss += loss.item()*data.shape[0]
                _,outputs = torch.max(outputs,1)
                running_correct += torch.sum(outputs == target.data)
            epoch_loss = running_loss / len(data_loader[phase].dataset)
            epoch_acc = running_correct.double() / len(data_loader[phase].dataset)
            
            if phase == 'val' and best_acc < epoch_acc:
                best_acc = epoch_acc
                best_epoch = epoch
                best_model = copy.deepcopy(model)

            print('[Epoch: %d] [%s] [loss: %.4f] [acc: %.4f]' % (epoch + 1,phase,epoch_loss,epoch_acc))
        end_time = time.time()
        inference_time = end_time - start_time
        print(f"Inference Time: {inference_time:.6f} seconds")

**Comment by xurui-joei** [comment]
> 作者大大您好，请问您计算Inference time是计算哪一部分的时间呢，我计算PU（参数一致）一个epoch的时间得到时间是0.81s，和论文差别较大。以下是代码
> 
> ```python
> ```     for epoch in range(epochs):
>         start_time = time.time()
>         for phase in ['train','val']:
>             running_loss = 0
>             running_correct = 0
>             for i, (data, target) in enumerate(data_loader[phase]):
>                 data, target = data.to(device), target.to(device)
>                 if phase == 'train':
>                     model.train()
>                     outputs = model(data)
>                     loss = criterion(outputs, target)
>                     optimizer.zero_grad()
>                     loss.backward()
>                     optimizer.step()
>                     scheduler.step()
>                 else:
>                     model.eval()
>                     with torch.no_grad():
>                         outputs = model(data)
>                         loss = criterion(outputs, target)
>                 running_loss += loss.item()*data.shape[0]
>                 _,outputs = torch.max(outputs,1)
>                 running_correct += torch.sum(outputs == target.data)
>             epoch_loss = running_loss / len(data_loader[phase].dataset)
>             epoch_acc = running_correct.double() / len(data_loader[phase].dataset)
>             
>             if phase == 'val' and best_acc < epoch_acc:
>                 best_acc = epoch_acc
>                 best_epoch = epoch
>                 best_model = copy.deepcopy(model)
> 
>             print('[Epoch: %d] [%s] [loss: %.4f] [acc: %.4f]' % (epoch + 1,phase,epoch_loss,epoch_acc))
>         end_time = time.time()
>         inference_time = end_time - start_time
>         print(f"Inference Time: {inference_time:.6f} seconds")
> ```

你好，论文里的inference time是test数据推理时间。

**Comment by machao0919** [comment]
作者大大请问您可以提供一下画分类图的这一小部分的代码吗，我想要将您的模型作为我对比实验，但是自己写的代码一直报错，调试了几天也没成功。

**Comment by xurui-joei** [comment]
> 作者大大请问您可以提供一下画分类图的这一小部分的代码吗，我想要将您的模型作为我对比实验，但是自己写的代码一直报错，调试了几天也没成功。

def plot(pred,target,dataset,model,train_ratio):
    h,w = target.shape
    cls_map = pred.reshape((h,w)) + 1
    spectral.save_rgb(f'path/to/pic/{dataset}_{train_ratio}_{model}.jpg',cls_map, colors=spectral.spy_colors) #使用绝对路径

all_pred,_ = test(device,model,all_loader)

plot(all_pred,gt,dataset_name,model_name,train_ratio)

你好，可以试一下上述代码

**Comment by xurui-joei** [comment]
> > 作者大大请问您可以提供一下画分类图的这一小部分的代码吗，我想要将您的模型作为我对比实验，但是自己写的代码一直报错，调试了几天也没成功。
> 
> ```
> def plot(pred,target,dataset,model,train_ratio):
>     h,w = target.shape
>     cls_map = pred.reshape((h,w)) + 1
>     #使用绝对路径
>     spectral.save_rgb(f'path/to/pic/{dataset}_{train_ratio}_{model}.jpg',cls_map, colors=spectral.spy_colors) 
> #____________________________________________________________________
> #下面代码接在train_test.py的最后
> all_pred,_ = test(device,model,all_loader)
> 
> plot(all_pred,gt,dataset_name,model_name,train_ratio)
> ```
> 
> 你好，可以试一下上述代码
